### ISSUE TITLE
Cache resolution of p2 queries

### ISSUE THREAD

**Issue by nedtwigg** [equodev/equo-ide#52]
Tasks can't add a maven dependency, so we sometimes need to do p2 queries at configuration time, which is bad. But, we can and should cache them, which would make things a *lot* faster.

- Add a new `public static File p2queries()` to 
  - https://github.com/equodev/equo-ide/blob/main/solstice/src/main/java/dev/equo/solstice/p2/CacheLocations.java
  - Rather than calculating the same queries over and over, we can record all the inputs to a query and quickly read the outputs.
- Create a new `class QueryCache`, it should have a constructor pretty much the same as this
  - https://github.com/equodev/equo-ide/blob/4844aa33e153943b8a936651679c45e743f3c1b5/solstice/src/main/java/dev/equo/solstice/p2/OfflineCache.java#L28-L39
- `P2Model` should implement `.equals`, `.hashCode` and `Serializable`
  - which means `Filter` needs to implement that too
- The overall shape of the cache is this:
  - Users have a `P2Model`, which is the key to the cache.
  - They call `P2Model.query()` which gives them a `P2Query`
  - https://github.com/equodev/equo-ide/blob/4844aa33e153943b8a936651679c45e743f3c1b5/solstice/src/main/java/dev/equo/solstice/p2/P2Model.java#L92
  - `P2Query` has a lot of data, but ultimately the only thing that really matters are these two method
    - https://github.com/equodev/equo-ide/blob/4844aa33e153943b8a936651679c45e743f3c1b5/solstice/src/main/java/dev/equo/solstice/p2/P2Query.java#L228
    - https://github.com/equodev/equo-ide/blob/4844aa33e153943b8a936651679c45e743f3c1b5/solstice/src/main/java/dev/equo/solstice/p2/P2Query.java#L240
- Here's an example of how a query gets made, some are added as maven dependencies, others are added as File dependencies downloaded by our P2 Client.
  - https://github.com/equodev/equo-ide/blob/4844aa33e153943b8a936651679c45e743f3c1b5/plugin-gradle/src/main/java/dev/equo/ide/gradle/P2DepsExtension.java#L51-L62
- I would define a class along these lines:

```java
public class P2QueryResult {
  List<String> mavenCoordinates;
  List<File> downloadedP2Jars;
  // (constructor.& getters)


  public static P2QueryResult cacheIfPossible(P2Model model, P2Client.Caching caching) {
    // if the P2Model is in the cache already, just use that
    // if it isn't, do the query, download all the jars, and put it in the cache
  }
}

```

**Comment by nedtwigg** [comment]
Also here

https://github.com/equodev/equo-ide/blob/a07fc0591a57d3e1931afaeba779b225c6fba974/plugin-gradle/src/main/java/dev/equo/ide/gradle/P2DepsExtension.java#L51-L62

**Comment by nedtwigg** [comment]
TODO: I will add more info to this later today.

**Comment by nedtwigg** [comment]
- [ ] if `url.toLowerCase().contains('snapshot') || ('latest')` don't do caching

**Comment by nedtwigg** [comment]
@maquinas07 FYI I just merged #83 which might have created minor merge conflicts for you. I recommend merging `main` now and working from there. I think we'll be working on different files from here on.
